hbase-server/src/main/java/org/apache/hadoop/hbase/wal/AbstractRecoveredEditsOutputSink.java

```diff
@@ -0,0 +1,283 @@
+      // writes the data without touching disk.
+      // TestHLogSplit#testThreading is an example.
+      if (walFS.exists(wap.path)) {
+        if (!walFS.rename(wap.path, dst)) {
+          throw new IOException("Failed renaming recovered edits " + wap.path + " to " + dst);
+        }
+        LOG.info("Rename recovered edits {} to {}", wap.path, dst);
+      }
+    } catch (IOException ioe) {
+      LOG.error("Could not rename recovered edits {} to {}", wap.path, dst, ioe);
+      thrown.add(ioe);
+      return null;
+    }
+    return dst;
+  }
+
+  @Override
+  public boolean keepRegionEvent(WAL.Entry entry) {
+    ArrayList<Cell> cells = entry.getEdit().getCells();
+    for (Cell cell : cells) {
+      if (WALEdit.isCompactionMarker(cell)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Update region's maximum edit log SeqNum.
+   */
+  void updateRegionMaximumEditLogSeqNum(WAL.Entry entry) {
+    synchronized (regionMaximumEditLogSeqNum) {
+      String regionName = Bytes.toString(entry.getKey().getEncodedRegionName());
+      Long currentMaxSeqNum = regionMaximumEditLogSeqNum.get(regionName);
+      if (currentMaxSeqNum == null || entry.getKey().getSequenceId() > currentMaxSeqNum) {
+        regionMaximumEditLogSeqNum.put(regionName, entry.getKey().getSequenceId());
+      }
+    }
+  }
+
+  // delete the one with fewer wal entries
+  private void deleteOneWithFewerEntries(RecoveredEditsWriter wap, Path dst)
+      throws IOException {
+    long dstMinLogSeqNum = -1L;
+    try (WAL.Reader reader = walSplitter.getWalFactory().createReader(walSplitter.walFS, dst)) {
+      WAL.Entry entry = reader.next();
+      if (entry != null) {
+        dstMinLogSeqNum = entry.getKey().getSequenceId();
+      }
+    } catch (EOFException e) {
+      LOG.debug("Got EOF when reading first WAL entry from {}, an empty or broken WAL file?", dst,
+          e);
+    }
+    if (wap.minLogSeqNum < dstMinLogSeqNum) {
+      LOG.warn("Found existing old edits file. It could be the result of a previous failed"
+          + " split attempt or we have duplicated wal entries. Deleting " + dst + ", length="
+          + walFS.getFileStatus(dst).getLen());
+      if (!walFS.delete(dst, false)) {
+        LOG.warn("Failed deleting of old {}", dst);
+        throw new IOException("Failed deleting of old " + dst);
+      }
+    } else {
+      LOG.warn("Found existing old edits file and we have less entries. Deleting " + wap.path
+          + ", length=" + walFS.getFileStatus(wap.path).getLen());
+      if (!walFS.delete(wap.path, false)) {
+        LOG.warn("Failed deleting of {}", wap.path);
+        throw new IOException("Failed deleting of " + wap.path);
+      }
+    }
+  }
+
+  /**
+   * Private data structure that wraps a {@link WALProvider.Writer} and its Path, also collecting
+   * statistics about the data written to this output.
+   */
+  final class RecoveredEditsWriter {
```

Comment:
Good.

Can it be static class?

Reply:
No need be static? It need to access the not-static filed and method in OutputSink.